### PR TITLE
564 automatic detection of include directives in memilio generation

### LINKED ISSUE #564: Automatic detection of include directives in memilio-generation
Memilio-generation doesn't generate the include directives. For now comments are placed as a placeholder and the user has to write them.

Now they are generated by informations retrieved during the generation. It uses information like which classes are used for the model and paths retrieved from the AST.

### PR DESCRIPTION
**Pull request by MaxBetzDLR**
## Changes

Add include directives to the automatic generation of pybind files.

## Merge Request - GuideLine Checklist

- *Check our [git workflow](https://github.com/DLR-SC/memilio/wiki/git-workflow) before opening a Pull request/Merge request.*
- *Request a reviewer when your work is ready to review, before this please use the draft feature.*

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/DLR-SC/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added

### Checks by code reviewer(s)

- [x] Corresponding issue(s) is/are linked and addressed
- [x] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [x] Appropriate **unit tests** have been added, CI passes and code coverage is acceptable (did not decrease)
- [x] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)

### Additional context

Please list additional information or things a reviewer should look out for.

Closes #564

### CODE REVIEW

**Comment by mknaranja** [comment]
@MaxBetzDLR Not sure this is finished. Could you ask for a reviewer once you want me to have a look?

**Comment by MaxBetzDLR** [comment]
> @MaxBetzDLR Not sure this is finished. Could you ask for a reviewer once you want me to have a look?

Issue should be finished know and is ready to be reviewed

File: pycode/memilio-generation/memilio/generation/template/template_string.py [hunk 1]
```diff
@@ -35,14 +35,37 @@ def includes(intermed_repr: IntermediateRepresentation) -> str:
...
-        "//Includes from Memilio\n"
-        "//Optional Includes: \"pybind11/stl_bind.h\", \"Eigen/Core\"\n"
+        "#include \"pybind_util.h\"\n"
+        "#include \"utils/custom_index_array.h\"\n"
+        "#include \"utils/parameter_set.h\"\n"
+        "#include \"utils/index.h\"\n"
     )
+
+    if "CompartmentalModel" in intermed_repr.model_base[0]:
+        substition_string += (
```
Comment (mknaranja) [inline pycode/memilio-generation/memilio/generation/template/template_string.py:46]: Probably all `substition_string ` should be replace by `substitution_string`

**Review by mknaranja — COMMENTED** [review]
Thank you @MaxBetzDLR. This looks clean and almost ready too merge. I still want to make another run with debug environment. However I again face problems with the directory derivation. Maybe we can have a look at that again. Otherwise it would be nice if you could correct the some typos I annotated.

File: pycode/memilio-generation/memilio/tools/config.json.txt [hunk 2]
```diff
@@ -14,7 +14,7 @@
         }
     },
     {
-        "source_file": "${memilio_SOURCE_DIR}/models/ode_seir/model.cpp",
+        "source_file": "${memilio_SOURCE_DIR}/models/ode_secir/model.cpp",
```
[the remark this answers is not in the document]
Comment (mknaranja) [reply pycode/memilio-generation/memilio/tools/config.json.txt:17]: Again, I cannot reproduce it. Maybe in between setting and resetting I forgot to load llvm or clang, I don't know. It's okay now and I will resolve this here.

**Comment by mknaranja** [comment]
@MaxBetzDLR Thanks for your work. I just corrected the secir namespace right now and will merge it once the CI has passed.

**Review by mknaranja — APPROVED** [review]

**Pull request merged**